iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_types.hpp

```diff
@@ -0,0 +1,98 @@
+#define IOX_HOOFS_MODULETESTS_TEST_CXX_FUNCTIONAL_INTERFACE_TYPES_HPP
+
+#include "test_cxx_functional_interface_common.hpp"
+
+namespace test_cxx_functional_interface
+{
+struct GenericValueErrorFactory
+{
+    using Type = GenericValueError;
+
+    using value_t = int;
+    using error_t = int;
+
+    static int usedTestValue;
+    static int anotherTestValue;
+
+    static int usedErrorValue;
+    static int anotherErrorValue;
+
+    static void configureNextTestCase() noexcept;
+    static GenericValueError createValidObject() noexcept;
+    static GenericValueError createInvalidObject() noexcept;
+};
+
+struct GenericPlainFactory
+{
+    using Type = GenericPlain;
+
+    static void configureNextTestCase() noexcept;
+    static Type createValidObject() noexcept;
+    static Type createInvalidObject() noexcept;
+};
+
+/// @brief Add here a type which inherits from FunctionalInterface and should
+///        be tested. Please consider GenericValueErrorFactory and GenericPlainFactory
+///        as a template.
+///
+///     Nullable class:
+///        If the class is just nullable but does not contain a value or an error
+///        then you have to create a struct like GenericPlainFactory with the methods
+///        * using Type =;
+///            type alias for the type which will be tested
+///
+///        * static void configureNextTestCase();
+///            called before every test case, can be useful to bring some randomisation into the
+///            createValidObject/createInvalidObject process.
+///
+///        * static Type createValidObject();
+///            creates a valid object (the operator bool has to return true, required for and_then case)
+///
+///        * static Type createInvalidObject();
+///            creates an invalid object (the operator bool has to return false, required for or_else case)
+///
+///     Class with value:
+///        A class with a value method requires additionally:
+///        * using value_t = ;
+///            Type alias of the value type
+///
+///        * static value_t usedTestValue;
+///            Value which was used while creating a valid object
+///
+///        * static value_t anotherTestValue;
+///            Another value which can be compared to usedTestValue and is not equal to it
+///
+///     Class with error:
+///        A class with a get_error method requires additionally:
+///        * using error_t = ;
+///            Type alias of the error type
+///
+///        * static error_t usedErrorValue;
+///            Error value which was used while creating an invalid object
+///
+///        * static error_t anotherErrorValue
+///            Another error value which can be compared to usedErrorValue and is not equal to it
+
+using FunctionalInterfaceImplementations = testing::Types<GenericValueErrorFactory, GenericPlainFactory>;
+TYPED_TEST_SUITE(FunctionalInterface_test, FunctionalInterfaceImplementations);
+
```

Comment:
What I'm missing is a check that the expected functional interface is implemented, e.g. that `cxx::optional` has `and_then` with value and `or_else` without value and not the other way around.
In the end, this will catch ones eye when the `optional` is used but that is also the case for many other things we test and since there is quite some template magic involved I think we should ensure that the expected interfaces are implemented. I think it might even be possible to test whether the wrong interfaces are not implemented. This should be doable by adding some `constexpr bool EXPECT_AND_THEN_WITH_VALUE {true}` or similar to the factory class.